gk/fib.c

```diff
@@ -0,0 +1,1337 @@
+				iface->name, __func__);
+
+		ret = setup_neighbor_tbl(socket_id, (identifier * 2),
+			ETHER_TYPE_IPv4, (1 << (32 - ip4_prefix_len)),
+			&fib[0]->u.neigh);
+		if (ret < 0) {
+			rte_lpm_delete(gk_conf->lpm_tbl.lpm,
+				ntohl(addr4.ip.v4.s_addr), ip4_prefix_len);
+			return -1;
+		}
+	}
+
+	/* Set up the FIB entry for the IPv6 network prefix. */
+	if (ipv6_if_configured(iface)) {
+		addr6.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&addr6.ip.v6,
+			&iface->ip6_addr, sizeof(addr6.ip.v6));
+		ip6_prefix_len = iface->ip6_addr_plen;
+
+		fib[1] = add_prefix_fib(&addr6, ip6_prefix_len, gk_conf);
+		if (fib[1] == NULL) {
+			ret = -1;
+			goto free_fib0;
+		}
+
+		if (iface == &net_conf->front) {
+			fib[1]->action = GK_FWD_NEIGHBOR_FRONT_NET;
+			gk_conf->neigh6_fib_front = fib[1];
+		} else if (likely(iface == &net_conf->back)) {
+			fib[1]->action = GK_FWD_NEIGHBOR_BACK_NET;
+			gk_conf->neigh6_fib_back = fib[1];
+		} else
+			rte_panic("Unexpected condiction: invalid interface %s at %s!\n",
+				iface->name, __func__);
+
+		ret = setup_neighbor_tbl(socket_id, (identifier * 2 + 1),
+			ETHER_TYPE_IPv6, gk_conf->max_num_ipv6_neighbors,
+			&fib[1]->u.neigh6);
+		if (ret < 0) {
+			rte_lpm6_delete(gk_conf->lpm_tbl.lpm6,
+				addr6.ip.v6.s6_addr, ip6_prefix_len);
+			goto free_fib0;
+		}
+	}
+
+	return 0;
+
+free_fib0:
+	if (fib[0] != NULL) {
+		rte_lpm_delete(gk_conf->lpm_tbl.lpm,
+			ntohl(addr4.ip.v4.s_addr), ip4_prefix_len);
+		destroy_neigh_hash_table(&fib[0]->u.neigh);
+	}
+
+	return -1;
+}
+
+static int
+init_fib_tbl(struct gk_config *gk_conf)
+{
+	int i;
+	int ret;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	rte_spinlock_init(&ltbl->lock);
+
+	for (i = 0; i < gk_conf->gk_max_num_fib_entries; i++) {
+		ltbl->fib_tbl[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl[i].num_updated_instances);
+
+		ltbl->fib_tbl6[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl6[i].num_updated_instances);
+	}
+
+	/* Set up the FIB entry for the front network prefixes. */
+	ret = setup_net_prefix_fib(0, &gk_conf->net->front, gk_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to setup the FIB entry for the front network prefixes at %s\n",
+			__func__);
+		goto out;
+	}
+
+	/* Set up the FIB entry for the back network prefixes. */
+	if (gk_conf->net->back_iface_enabled) {
+		ret = setup_net_prefix_fib(1, &gk_conf->net->back, gk_conf);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: failed to setup the FIB entry for the back network prefixes at %s\n",
+				__func__);
+			goto free_front_fibs;
+		}
+	}
+
+	ret = 0;
+	goto out;
+
+free_front_fibs:
+	if (gk_conf->neigh_fib_front != NULL)
+		destroy_neigh_hash_table(&gk_conf->neigh_fib_front->u.neigh);
+	if (gk_conf->neigh6_fib_front != NULL)
+		destroy_neigh_hash_table(&gk_conf->neigh6_fib_front->u.neigh6);
+
+out:
+	return ret;
+}
+
+/*
+ * XXX Only instantiate the LPM tables needed, for example,
+ * there's no need for an IPv6 LPM table in an IPv4-only deployment.
```

Comment:
I think this patch is the only piece of code in Gatekeeper that doesn't support IPv4-only and IPv6-only deployments. If so, it's better to take on the extra work and implement this `XXX` comment out.